riscv_isac/coverage.py

```diff
@@ -957,17 +963,92 @@ def old_fn_csr_comb_covpt(csr_reg):
             for i in csr_regfile.csr_regs:
                 instr_vars[i] = int(csr_regfile[i],16)
 
+            instr.iptw_update(instr_vars, iptw_dict)
+            instr.ptw_update(instr_vars)
+
+            for i in iptw_dict:
+                instr_vars[i] = (iptw_dict[i])
+
             csr_write_vals = {}
             if instr.csr_commit is not None:
                 for commit in instr.csr_commit:
-                    if commit[0] == "CSR" and commit[3]:
-                        csr_write_vals[commit[1]] = int(commit[3],16)
+                    if commit[0] == "CSR" and commit[4] and commit[2] == '<-':
+                        csr_write_vals[commit[1]] = int(commit[4],16)
             def write_fn_csr_comb_covpt(csr_reg):
                 if csr_reg in csr_write_vals:
                     return csr_write_vals[csr_reg]
                 else:
                     return int(csr_regfile[csr_reg],16)
+                
+            csr_read_vals = {}
+            if instr.csr_commit is not None:
+                for csr_commit in instr.csr_commit:
+                    if(csr_commit[0] == "CSR") and (csr_commit[3]) and (csr_commit[2] == '->'):
+                        csr_read_vals[csr_commit[1]] = int(csr_commit[3],16)
+            def read_fn_csr_comb_covpt(csr_reg):
+                if instr.csr_commit is not None and csr_reg in csr_read_vals:
+                    return csr_read_vals.get(csr_reg)
+                else:
+                    return None
+
+            def check_label_address(label):
+                return utils.collect_label_address(elf, label)
+
+            def get_mem_val(addr, bytes):
+                if addr in mem_vals:
+                    return mem_vals.get(addr)
+                else:
+                    mem_val = utils.get_value_at_location(elf_path = elf, location = addr, bytes = bytes)
```

Comment:
The value of the data should never be fetched from the elf as the elf would only contain the initial value. If a value is being used in the coverpoints, it should be obtained only from the logs. If the logs do not contain the value, they need to be fixed. Getting it from the elf might lead to false positives in coverage if the implementation is not correct.

Reply:
If you please refer to **c_sail.py**, the function for extracting the values at every memory address from the log file for each instruction has been implemented. So, the flow of `mem_val` is such that it will first get the values of every address from the log files, and then, here in the **coverage.py** when the function `get_mem_val` is used, it first checks the `mem_val` and if this dictionary does not contain the value of that specific address mentioned in the coverpoint which will only happen if that address was initialized in the test(.S) but never used. Therefore, we will have the track of every address value whether updated or not by the log(s).

Reply:
My interpretation is that this is just to ensure that statically initialized data gets initialized correctly. Do you see a problem with it?

Reply:
While, I see the value in ensuring that the statically initialized data is checking correctly, if the value is not loaded into the simulator properly i.e the (.S) files are not representative of the actual value loaded in SAIL we risk false positives. If the logs always contain the value being read(I believe this is already the case), I don't think there is any issue with not supporting this feature.